tools/mutation_test/mutator.py

```diff
@@ -0,0 +1,499 @@
+
+        For each pattern, the list contains a maximum of `mutants_per_pattern` number of 
+        occurrences. Default is 10 if not provided.
+
+        If `random` is True, then the list is shuffled using this mutator's rng
+        """
+        utils.yellow_print("Searching for patterns...")
+        # list of mutations to execute, contains occurrence objects
+        mutations_list = []
+        # dictionary that maps a pattern to the list of occurrences
+        mutations_dict = {}
+        # go through each pattern
+        for mp in self.getPatterns():
+            # find their occurrences
+            occurrences_with_mp = self.findOccurrences(mutation_pattern=mp)
+            # for o in occurrences_with_mp:
+            #     utils.yellow_print(o)
+            # if occurrences are found add them to dictionary
+            if mp not in mutations_dict:
+                mutations_dict[mp] = []
+            if random:
+                self.rng.shuffle(occurrences_with_mp)
+            mutations_dict[mp] += occurrences_with_mp[0:mutants_per_pattern]
+        for mp in mutations_dict:
+            mutations_list += mutations_dict[mp]
+        if random:
+            self.rng.shuffle(mutations_list)
+        return mutations_list
+    
+    def nextRandomPattern(self):
+        """ Gets a random pattern
+
+        Ensures that the same pattern is not selected twice
+        """
+        if len(self.pattern_generator) == 0:
+            self.pattern_generator = self.mutation_patterns.copy()
+        index = self.rng.randint(0, len(self.pattern_generator) - 1)
+        ret = self.pattern_generator[index]
+        self.pattern_generator[index] = self.pattern_generator[len(self.pattern_generator) - 1]
+        self.pattern_generator[len(self.pattern_generator) - 1] = ret
+        self.pattern_generator.pop()
+        return ret
+
+    def generateRandomMutant(self) -> Tuple[Pattern, Occurrence, str, str]:
+        """ Generates a mutant by choosing random pattern then a random line
+
+        Not recommended to use everytime, it is a better idea to set up your own
+        runs with the other API methods.
+
+        This can be used as an example.
+        """
+        # Reset pattern_generator to all patterns
+        self.pattern_generator = self.mutation_patterns.copy()
+
+        # Go through every pattern to find an occurrence
+        while len(self.pattern_generator) > 0:
+            pattern = self.nextRandomPattern()
+            occurrences = self.findOccurrences(pattern)
+            if len(occurrences) != 0:
+                occurrence = self.rng.choice(occurrences)
+                original_line, mutated_line = self.mutate(occurrence)
+                return pattern, occurrence, original_line, mutated_line
+        sys.stderr.write("Could not create a mutant. Please make sure it is a C file.\n")
+        sys.stderr.write("You may need to indent your C file.\n")
+        return None
+
```

Comment:
these trailing newline chars can be removed

Reply:
https://github.com/aws/amazon-freertos/pull/2475/commits/ad49bdec868f800f447b238ed16082ee00c4209b